tests/unittests/test_devmode.py

```diff
@@ -0,0 +1,105 @@
+        return os.path.join(self.tmpdir, self.predictable_filename)
+
+    @patch("requests.Session.post", side_effect=mocked_auth_post)
+    def test_client_without_dev_mode(self, mocked_auth_post):
+        """checks if the client can write refresh token and expiry to
+        config."""
+
+        with open(self.tmp_config_path, "w") as ff:
+            json.dump(self.base_config, ff)
+
+
+        client = EloquaClient(config_path=self.tmp_config_path,dev_mode=False,config=self.base_config)
+        client.get_access_token()
+        with open(self.tmp_config_path) as config_file:
+            config = json.load(config_file)
+            self.assertEqual(True, "expires_in" in config)
+            self.assertEqual(True, "access_token" in config)
+            self.assertEqual(True, "refresh_token" in config)
+
+    def test_devmode_accesstoken_absent(self, *args, **kwargs):
+        """checks exception if access token is not present in config."""
+
+        with open(self.tmp_config_path, "w") as config_file:
+            json.dump(self.base_config, config_file)
+
+        try:
+            client = EloquaClient(config_path=self.tmp_config_path,dev_mode=True,config=self.base_config)
+
+            client.get_access_token()
+        except Exception as error:
+            self.assertEqual(str(error), "Unable to locate key in config")
+
+    @patch("requests.Session.post", side_effect=mocked_auth_post)
+    def test_client_valid_token(self, mocked_auth_post):
+        """checks if the client can fetch and validate refresh token and expiry
+        from config."""
+
+        config_sample = {
+            "access_token": "token",
+            "expires_in": strftime(datetime.datetime.utcnow().replace(tzinfo=pytz.UTC) + datetime.timedelta(days=5)),
+            **self.base_config,
+        }
+        with open(self.tmp_config_path, "w") as config_file:
+            json.dump(config_sample, config_file)
+        client = EloquaClient(config_path=self.tmp_config_path,dev_mode=True,config=config_sample)
+        client.get_access_token()
+
+    @patch("requests.Session.post", side_effect=mocked_auth_post)
+    def test_client_invalid_token(self, mocked_auth_post):
+        """checks if the client can fetch and validate refresh token and expiry
+        from config."""
+        expiry_time = strftime(datetime.datetime.utcnow().replace(tzinfo=pytz.UTC) - datetime.timedelta(minutes=5))
+        config_sample = {
+            "access_token": "token",
+            "expires_in": expiry_time,
+            **self.base_config,
+        }
+        with open(self.tmp_config_path, "w") as config_file:
+            json.dump(config_sample, config_file)
+        try:
+            client = EloquaClient(config_path=self.tmp_config_path,dev_mode=True,config=config_sample)
+            client.get_access_token()
+        except Exception as error:
+            self.assertEqual(str(error), "Access Token in config is expired, unable to authenticate in dev mode")
```

Comment:
Use with self.assertRaises() unittest construct here.

Reply:
I can't use assertRaises here, because we can only check the "type" of the exception with it.
But in this case, all exceptions are raised with "Exception" class and I want to validate the message-string that is raised with it.

Reply:
[Docs for `assertRaises`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises). These docs show how you can use a context manager with `assertRaises`. And they show that the exception object is a property on the manager, which means we can do exactly what you want to do

Here's how you use it
```python
$ python3
Python 3.10.9 (main, Dec 15 2022, 18:25:35) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.

# start setup

>>> import unittest
>>> class some_test(unittest.TestCase):
...   pass
...
>>> my_test = some_test()


# Prove assertRaises works

>>> with my_test.assertRaises(Exception):
...   1+1
...
2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 226, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/usr/local/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/unittest/case.py", line 163, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: Exception not raised


# Prove the test can pass

>>> with my_test.assertRaises(Exception) as context_manager:
...   raise Exception("this is a test")
...


# Show we have the exception

>>> actual_exception = context_manager.exception
>>> actual_exception
Exception('this is a test')
>>> str(actual_exception)
'this is a test'
```